modules/tensor_mechanics/src/materials/ADViscoplasticityStressUpdate.C

```diff
@@ -0,0 +1,290 @@
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::computeResidual(const ADReal & equiv_stress,
+                                                              const ADReal & trial_gauge)
+{
+  const ADReal M = std::abs(_hydro_stress) / trial_gauge;
+  const ADReal dM_dtrial_gauge = -M / trial_gauge;
+
+  const ADReal residual_left = Utility::pow<2>(equiv_stress / trial_gauge);
+  const ADReal dresidual_left_dtrial_gauge = -2.0 * residual_left / trial_gauge;
+
+  ADReal residual = residual_left;
+  _derivative = dresidual_left_dtrial_gauge;
+
+  if (_pore_shape == PoreShapeModel::SPHERICAL)
+  {
+    residual *= 1.0 + _intermediate_porosity / 1.5;
+    _derivative *= 1.0 + _intermediate_porosity / 1.5;
+  }
+
+  if (_model == ViscoplasticityModel::GTN)
+  {
+    residual += 2.0 * _intermediate_porosity * std::cosh(_pore_shape_factor * M) - 1.0 -
+                Utility::pow<2>(_intermediate_porosity);
+    _derivative += 2.0 * _intermediate_porosity * std::sinh(_pore_shape_factor * M) *
+                   _pore_shape_factor * dM_dtrial_gauge;
+  }
+  else
+  {
+    const ADReal h = computeH(_power, M);
+    const ADReal dh_dM = computeH(_power, M, true);
+
+    residual += _intermediate_porosity * (h + _power_factor / h) - 1.0 -
+                _power_factor * Utility::pow<2>(_intermediate_porosity);
+    const ADReal dresidual_dh = _intermediate_porosity * (1.0 - _power_factor / Utility::pow<2>(h));
+    _derivative += dresidual_dh * dh_dM * dM_dtrial_gauge;
+  }
+
+  if (_verbose)
+  {
+    Moose::out << "in computeResidual:\n"
+               << "  position: " << _q_point[_qp] << " hydro_stress: " << _hydro_stress
+               << " equiv_stress: " << equiv_stress << " trial_grage: " << trial_gauge
+               << " M: " << M << std::endl;
+    Moose::out << "  residual: " << residual << "  derivative: " << _derivative << std::endl;
+  }
+
+  return residual;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::computeH(const Real n,
+                                                       const ADReal & M,
+                                                       const bool derivative)
+{
+  const ADReal mod = std::pow(M * _pore_shape_factor, (n + 1.0) / n);
+
+  // Calculate derivative with respect to M
+  if (derivative)
+  {
+    const ADReal dmod_dM = (n + 1.0) / n * mod / M;
+    return dmod_dM * std::pow(1.0 + mod / n, n - 1.0);
+  }
+
+  return std::pow(1.0 + mod / n, n);
+}
+
+template <ComputeStage compute_stage>
+ADRankTwoTensor
+ADViscoplasticityStressUpdate<compute_stage>::computeDGaugeDSigma(
+    const ADReal & gauge_stress,
+    const ADReal & equiv_stress,
+    const ADRankTwoTensor & dev_stress,
+    const ADRankTwoTensor & stress)
+{
+  // Compute the derivative of the gauge stress with respect to the equilvalent and hydrostatic
+  // stress components
+  const ADReal M = std::abs(_hydro_stress) / gauge_stress;
```

Comment:
Do you need to guard against division by zero here and at various other places in this method?

Generally, i don't like having global variables like _hydro_stress, _intermediate_porosity, etc, floating around.  This comes from my fortran days where global vars were a real nightmare because you couldn't see what was changing them.   You probably toyed with not having such global vars and decided they were OK.  But if not, i think it'd be worth trying to get rid of them.

Reply:
I tried to catch all the divide by zeros. A lot of this is protected by limiting the code that is run if there is no equivalent stress, and forcing gauge stress to be >= equivalent stress.

As for global variables, they're somewhat necessary due to the code inheritance with limited parameters passed to the methods, especially those from ADSingleVariableReturnMappingSolution.